Add CLI App Support

kschwab opened the pull request:
Adds utility functions for running CLI applications.

See [documentation](https://github.com/kschwab/pydantic-settings/blob/b47e1647fc067ca51f8a50abdd41568f72fc6282/docs/index.md#creating-cli-applications) for details.

kschwab commented on docs/index.md:
```diff
@@ -560,19 +559,7 @@ print(Settings().model_dump())
-    this_foo: str
-
-
-print(Settings(_cli_parse_args=['--this_foo', 'is such a foo']).model_dump())
```
Fixed suggestion of using private param `_cli_parse_args`. The "formal" equivalent is now:

```python
CliApp.run(Settings, cli_args=['--this_foo', 'is such a foo']).model_dump()
```

kschwab commented on docs/index.md:
```diff
@@ -1211,13 +1287,15 @@ cli_settings = CliSettingsSource(Settings, root_parser=parser)
 
 # Parse and load CLI settings from the command line into the settings source.
 sys.argv = ['example.py', '--food', 'kiwi', '--name', 'waldo']
-print(Settings(_cli_settings_source=cli_settings(args=True)).model_dump())
```
Fixed suggestions of using private param `_cli_settings_source` for its formal equivalent.

kschwab commented on pydantic_settings/main.py:
```diff
@@ -33,7 +40,6 @@ class SettingsConfigDict(ConfigDict, total=False):
     env_parse_enums: bool | None
     cli_prog_name: str | None
     cli_parse_args: bool | list[str] | tuple[str, ...] | None
-    cli_settings_source: CliSettingsSource[Any] | None
```
This was a bug. It's not possible to provide `cli_settings_source` as a configuration setting as the `settings_cls` cannot be defined (circular dependency).

hramezani commented:
Thanks @kschwab for this PR.

I have some concerns about the cli settings source, it has a lot of code and I think maintaining this amount of code would be hard in the future.

Specially about `CLI App`, Is it requested by somebody?

kschwab commented:
@hramezani, yes, this was requested and discussed in #335 and #336. It is related to discussion we had on [get_subcomand](https://github.com/pydantic/pydantic-settings/pull/341#issuecomment-2310615348), where this functionality for CLI is somewhat standard. I think drawing the line here for feature support makes sense.

Also, I understand the LOC concern for `CliSettingsSource`, however the root of this is in `sources.py`. With everything now in and understood, it can be simplified. I was planning on addressing this separately.

hramezani commented:
@kschwab Don't we need more tests for this patch? I can't see any test for `CliApp.run` in particular.  did I miss something?

kschwab commented:
@hramezani it's mostly a superficial commit. The doc tests cover happy paths, and I added tests in `test_settings.py` for exception paths. Would you like me to copy the doc tests into `test_settings.py`?

kschwab commented:
@hramezani, I added more tests specific to `CliApp.run` and updated all other test references using `_cli_parse_args=...` to instead use `CliApp.run(Settings, cli_args=...)` where relevant.

hramezani commented:
Thanks @kschwab.

I will check it next week

hramezani commented:
Thanks @kschwab
